ledger/eval.go

```diff
@@ -995,40 +981,36 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
 	}
 
 	var txvalidator evalTxValidator
-	ctx, cf := context.WithCancel(ctx)
-	defer cf()
+	validationCtx, validationCancel := context.WithCancel(ctx)
+	defer validationCancel()
 	if validate {
-		proto, ok := config.Consensus[blk.CurrentProtocol]
+		_, ok := config.Consensus[blk.CurrentProtocol]
 		if !ok {
 			return StateDelta{}, protocol.Error(blk.CurrentProtocol)
 		}
 		txvalidator.txcache = txcache
 		txvalidator.block = blk
-		txvalidator.proto = proto
 		txvalidator.verificationPool = executionPool
 
-		txvalidator.ctx = ctx
-		txvalidator.cf = cf
-		txvalidator.txgroups = make(chan []transactions.SignedTxnWithAD, len(paysetgroups))
+		txvalidator.ctx = validationCtx
+		txvalidator.txgroups = paysetgroups
 		txvalidator.done = make(chan error, 1)
 		go txvalidator.run()
+
 	}
 
 	for _, txgroup := range paysetgroups {
 		select {
 		case <-ctx.Done():
-			select {
-			case err := <-txvalidator.done:
+			return StateDelta{}, ctx.Err()
+		case err, open := <-txvalidator.done:
+			// if we're not validating, then `txvalidator.done` would be nil, in which case this case statement would never be executed.
+			if open && err != nil {
 				return StateDelta{}, err
-			default:
 			}
-			return StateDelta{}, ctx.Err()
 		default:
 		}
 
-		if validate {
-			txvalidator.txgroups <- txgroup
-		}
```

Comment:
another parallelism idea: run NumCPU-1 goroutines that equieue tasks received via `txvalidator.txgroups` channel. It possibly will have much less code but the same perf/parallelism.

Reply:
The execution pool is already using the NumCPU; that's why we use it..